src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -0,0 +1,374 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Net.Http.Headers;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace System.Net.Http
+{
+    internal partial class AuthenticationHelper
+    {
+        public const string Digest = "Digest";
+
+        // Define digest constants
+        private const string Qop = "qop";
+        private const string Auth = "auth";
+        private const string AuthInt = "auth-int";
+        private const string Nonce = "nonce";
+        private const string NC = "nc";
+        private const string Realm = "realm";
+        private const string UserHash = "userhash";
+        private const string Username = "username";
+        private const string Algorithm = "algorithm";
+        private const string Uri = "uri";
+        private const string Sha256 = "SHA-256";
+        private const string Md5 = "MD5";
+        private const string Sha256Sess = "SHA-256-sess";
+        private const string CNonce = "cnonce";
+        private const string Opaque = "opaque";
+        private const string Response = "response";
+        private const string Colon = ":";
+
+        // Define alphanumeric characters for cnonce
+        // 48='0', 65='A', 97='a'
+        private static int[] AlphaNumChooser = new int[] { 48, 65, 97 };
+
+        public static bool TrySetDigestAuthToken(HttpRequestMessage request, ICredentials credentials, DigestResponse digestResponse)
+        {
+            NetworkCredential credential = credentials.GetCredential(request.RequestUri, Digest);
+            if (credential == null)
+            {
+                return false;
+            }
+
+            try
+            {
+                request.Headers.Authorization = new AuthenticationHeaderValue(Digest, GetDigestTokenForCredential(credential, request, digestResponse));
+                return true;
+            }
+            catch
+            {
+                // Return false in case of any digest header calculation errors.
+                return false;
+            }
+        }
+
+        public static string GetDigestTokenForCredential(NetworkCredential credential, HttpRequestMessage request, DigestResponse digestResponse)
+        {
+            StringBuilder sb = StringBuilderCache.Acquire();
+
+            // It is mandatory for servers to implement sha-256
+            // Keep MD5 for backward compatibility.
+            string algorithm = Md5;
+            if (digestResponse.Parameters[Algorithm].Contains(Sha256))
+            {
+                algorithm = Sha256;
```

Comment:
The algorithm may not by SHA256 or MD5. A server could specify SHA512 for example and this code would fall back to MD5. The choice criteria should be, if none specified or MD5 specified, use MD5. If SHA256 specified, use SHA256. Otherwise fail authentication.